### PR TITLE
V3 runtime parser

### PR DESCRIPTION
**Pull request by zbraniecki**
@stasm - f?

### CODE REVIEW

**Comment by zbraniecki** [comment]
Stas, I think you can skip the changes to both parsers for now. I mostly would like to get your feedback to changes to resolver/pseudo etc. and maybe just scan through tests to see if there's any change that worries you. Thanks!

File: src/lib/env.js [hunk 1]
```diff
@@ -60,19 +60,20 @@ export class Env {
...
+    }
   }
 
   return entries;
 }
 
-function createPseudoEntry(node, lang) {
+function createPseudoEntry(id, node, lang) {
   return createEntry(
     walkContent(node, qps[lang.code].translate), lang);
```
Comment (stasm) [inline src/lib/env.js]: I think you'll need to pass `id` here as the first arg now?

File: src/lib/resolver.js [hunk 2]
```diff
@@ -15,53 +15,21 @@ var PDI = '\u2069';
...
 
-export function createEntry(node, lang) {
-  var keys = Object.keys(node);
+export function createEntry(id, entity, lang) {
+  meta.set(entity, {id, lang});
 
-  // the most common scenario: a simple string with no arguments
-  if (typeof node.$v === 'string' && keys.length === 2) {
-    return node.$v;
+  if (entity.attrs === undefined) {
```
Comment (stasm) [inline src/lib/resolver.js]: You might not need this check since `for...in` gracefully doesn't do anything for `undefined`.

**Comment by stasm** [comment]
I'll have a deeper look tomorrow morning.  Skimming the changes, I like the direction a lot.  Let's coordinate landing this with changes to master's AST and my plan to migrate the buildtime to l20n.js completely.

**Comment by zbraniecki** [comment]
FM, non-default (fr), Flame

master:
Metric                            Mean     Median   Min      Max       StdDev  p95     
--------------------------------  -------  -------  -------  --------  ------  --------
coldlaunch.navigationLoaded       517.100  517.000  467.000  569.000   31.063  568.500 
coldlaunch.navigationInteractive  750.800  726.500  670.000  893.000   67.703  883.500 
coldlaunch.visuallyLoaded         750.900  726.500  670.000  893.000   67.650  883.500 
coldlaunch.contentInteractive     751.400  726.500  671.000  894.000   67.621  884.500 
coldlaunch.fullyLoaded            897.250  877.500  819.000  1049.000  61.035  1033.500
coldlaunch.uss                    11.010   11.100   10.600   11.200    0.176   11.200  
coldlaunch.rss                    29.210   29.300   28.800   29.400    0.176   29.400  
coldlaunch.pss                    14.410   14.500   14.000   14.600    0.176   14.600  

v3:
Metric                            Mean     Median   Min      Max       StdDev  p95     
--------------------------------  -------  -------  -------  --------  ------  --------
coldlaunch.navigationLoaded       544.650  539.000  486.000  635.000   36.382  619.500 
coldlaunch.navigationInteractive  769.450  746.000  688.000  954.000   68.987  931.000 
coldlaunch.visuallyLoaded         770.050  746.500  689.000  954.000   68.861  931.000 
coldlaunch.contentInteractive     770.600  747.000  689.000  955.000   68.902  932.000 
coldlaunch.fullyLoaded            925.550  907.000  853.000  1110.000  62.862  1080.500
coldlaunch.pss                    14.510   14.500   14.000   14.800    0.184   14.750  
coldlaunch.rss                    29.325   29.300   28.800   29.600    0.197   29.600  
coldlaunch.uss                    11.105   11.100   10.600   11.400    0.183   11.350 

v3-runtime-parser:
Metric                            Mean     Median   Min      Max       StdDev  p95     
--------------------------------  -------  -------  -------  --------  ------  --------
coldlaunch.navigationLoaded       532.000  524.500  481.000  596.000   32.147  595.000 
coldlaunch.navigationInteractive  776.700  754.000  660.000  949.000   90.179  949.000 
coldlaunch.visuallyLoaded         776.750  754.000  660.000  950.000   90.275  949.500 
coldlaunch.contentInteractive     777.300  754.500  661.000  950.000   89.966  949.500 
coldlaunch.fullyLoaded            933.350  906.500  848.000  1120.000  78.149  1102.500
coldlaunch.pss                    14.410   14.450   13.900   14.800    0.281   14.800  
coldlaunch.uss                    11.015   11.050   10.500   11.400    0.283   11.400  
coldlaunch.rss                    29.220   29.250   28.800   29.600    0.264   29.600

File: src/lib/resolver.js [hunk 3]
```diff
@@ -15,53 +15,21 @@ var PDI = '\u2069';
 const meta = new WeakMap();
 const resolutionChain = new WeakSet();
 
-export function createEntry(node, lang) {
-  var keys = Object.keys(node);
+export function createEntry(id, entity, lang) {
+  meta.set(entity, {id, lang});
```
Comment (stasm) [inline src/lib/resolver.js]: Do we need to be setting this for entities which are simple strings?

Comment (stasm) [inline src/lib/resolver.js]: A way to check that, other than `typeof === 'string'`, would be to not pass `id` directly into `createEntry`, and instead check if `entity.id` is present.  Simple strings don't have the `id` defined on them.

Comment (stasm) [inline src/lib/resolver.js]: Ah, no, sorry, my bad.  We don't have `id` on entity anymore ^^.  Ignore the previous comment.

**Comment by stasm** [comment]
I added a _legacy_ mode in #48 which should help us migrate to l20n in Gaia buildtime and merge this in at the same time.  v2 AST will be handled by files in src/bindings/gaiabuild/legacy/*.

**Pull request closed without merging**